cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i

```diff
@@ -633,6 +712,244 @@ typename DataHelper::NodeType::Base * TrainBatchTaskBase<algorithmFPType, DataHe
+        {}
+
+        WorkItem & operator=(const WorkItem & src)
+        {
+            if (src.isLeaf)
+            {
+                improvement = 0.0;
+                isLeaf      = true;
+                return *this;
+            }
+
+            isLeaf           = src.isLeaf;
+            featureUnordered = src.featureUnordered;
+            start            = src.start;
+            n                = src.n;
+            nLeft            = src.nLeft;
+            level            = src.level;
+            improvement      = src.improvement;
+            impurityLeft     = src.impurityLeft;
+            impurityRight    = src.impurityRight;
+            node             = src.node;
+
+            return *this;
+        }
+    };
+
+    if (_hostApp.isCancelled(s, n))
+    {
+        return nullptr;
+    }
+    BinaryHeap<WorkItem, cpu> binaryHeap(s);
+    if (!s.ok())
+    {
+        return nullptr;
+    }
+    size_t remainingSplitNodes = _maxLeafNodes - 1;
+
+    // Create base node
+    typename DataHelper::TSplitData split;
+    IndexType iFeature;
+    WorkItem base(bUnorderedFeaturesUsed, iStart, n, level);
+
+    if (!remainingSplitNodes || terminateCriteria(n, level, curImpurity))
+    {
+        return makeLeaf(_aSample.get() + iStart, n, curImpurity, nClasses);
+    }
+    if (findBestSplit(iStart, n, curImpurity, iFeature, split))
+    {
+        const size_t nLeft   = split.nLeft;
+        const double imp     = curImpurity.var;
+        const double impLeft = split.left.var;
+
+        // check impurity decrease
+        double improvement = imp * n - impLeft * nLeft - (n - nLeft) * (imp - impLeft);
+        if (improvement < _minImpurityDecrease)
+        {
+            return makeLeaf(_aSample.get() + iStart, n, curImpurity, nClasses);
+        }
+        if (_par.varImportance == training::MDI)
+        {
+            addImpurityDecrease(iFeature, n, curImpurity, split);
+        }
+
+        base.nLeft        = split.nLeft;
+        base.improvement  = improvement;
+        base.impurityLeft = split.left;
+        _helper.convertLeftImpToRight(n, curImpurity, split);
+        base.impurityRight = split.left;
+        if (!(base.node = makeSplit(iFeature, split.featureValue, split.featureUnordered, nullptr, nullptr, curImpurity.var)))
+        {
+            return nullptr;
+        }
+        base.isLeaf = false;
+        base.featureUnordered |= bool(split.featureUnordered);
+        base.node->count = n;
+
+        --remainingSplitNodes;
+        s = binaryHeap.push(base);
+        if (!s.ok())
+        {
+            return nullptr;
+        }
+    }
+    else
+    {
+        return makeLeaf(_aSample.get() + iStart, n, curImpurity, nClasses);
+    }
+
+    while (!binaryHeap.empty())
+    {
+        WorkItem & src = binaryHeap.pop();
+        if (src.isLeaf)
+        {
+            continue;
+        }
+
+        // create leftChild
```

Comment:
I see 775-823, 833-879 and 885-935 lines a similar code (duplicate). we can generalize it. else it will be very difficult to support, optimize it, etc.

Reply:
agree, it was fixed